% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/mutCovariate.indel.compile.R
\name{mutCovariate.indel.compile}
\alias{mutCovariate.indel.compile}
\title{Compile covariates matrix for all chromosomes together.}
\usage{
mutCovariate.indel.compile(
  mask.regions.file = system.file("extdata", "mask_regions.RDS", package = "MutSpot"),
  all.sites.file = system.file("extdata", "all_sites.RDS", package = "MutSpot"),
  indel.mutations.file,
  sample.specific.features.url.file = NULL,
  region.of.interest,
  cores = 1,
  indel.mutations.file2,
  chrom.dir,
  genome.build = "Ch37"
)
}
\arguments{
\item{mask.regions.file}{Regions to mask in genome, for example, non-mappable regions/immunoglobin loci/CDS regions RDS file, depends on genome build, default file = mask_regions.RDS, Ch37.}

\item{all.sites.file}{All sites in whole genome RDS file, depends on genome build, default file = all_sites.RDS, Ch37.}

\item{indel.mutations.file}{Indel mutations found in region of interest MAF file.}

\item{sample.specific.features.url.file}{Text file containing sample specific indel features, default = NULL.}

\item{region.of.interest}{Region of interest bed file, default = NULL.}

\item{cores}{Number of cores, default = 1.}

\item{indel.mutations.file2}{Indel mutations MAF file.}

\item{chrom.dir}{Directory containing feature matrix of each chromosome.}

\item{genome.build}{Reference genome build, default = Ch37.}
}
\value{
Full covariates matrix for all chromosomes together.
}
\description{
Compile covariates matrix for all chromosomes together.
}
