---
title: "Supplementary Figure 6 - Performance of MutSpot on melanoma"
output: html_document
---

This is the R Markdown for Supplementary Figure 6, which consists of 1 part.

```{r, echo = FALSE}
suppressWarnings(suppressMessages(library(MutSpot)))
suppressWarnings(suppressMessages(library(gridExtra)))
suppressWarnings(suppressMessages(library(GenomicRanges)))
suppressWarnings(suppressMessages(library(ggplot2)))
suppressWarnings(suppressMessages(library(plotrix)))
```

## Figure A
Hotspots detected by MutSpot on melanoma samples
```{r, eval = FALSE}
MutSpot(snv.mutations = "melanoma_snv_mutations.MAF",
        cores = 6,
        fdr.cutoff = 0.01,
        min.count.snv = 4,
        hotspot.size = 31,
        drop = TRUE,
        cutoff.features.new.snv = 0.98,
        cutoff.nucleotide.new = 1,
        sample.snv.features ="sample_features_table_melanoma.txt",
        genomic.features = "genomic_features_melanoma_CTTCCG2_TF_prom.txt")
```

```{r}
hotspots.file = read.delim("snv_hotspots_merged_melanoma_min4_31bp.tsv", stringsAsFactors = FALSE)
fdr.cutoff = 0.01
color.line = "red"
color.dots = "maroon1"

# Adjust manhattan plot cutoff from fdr 0.1 to fdr 0.05, highlight proximal TFBS overlaps and save as SVG with height = 5 and width = 8
  hotspots.plot = hotspots.file
  hotspots.plot$region = rownames(hotspots.plot)
  
  x = hotspots.plot
  x$chrom = as.character(x$chrom)
  x$CHR = ifelse(x$chrom == "chrX", "23", substr(x$chrom, 4, nchar(x$chrom)))
  x$CHR = as.numeric(x$CHR)
  x$bp = ceiling((x$start + x$end) / 2)
  x$transcript_id = x$region
  
  highlight = x[which(x$fdr <= fdr.cutoff), "transcript_id"]
  if (length(highlight) == 0) {
    
    highlight = NULL
    fdr.cutoff = NULL
    
  } else {
    
    fdr.cutoff = mean(c(max(x[which(x$fdr <= fdr.cutoff), "pval"]), min(x[which(x$fdr > fdr.cutoff), "pval"])))
    
  }
  
  motifa = read.delim("CTTCCG2aprom.bed", stringsAsFactors = FALSE, header = FALSE)
  motifb = read.delim("CTTCCG2bprom.bed", stringsAsFactors = FALSE, header = FALSE)
  motifa$V2 = motifa$V2 + 1
  motifb$V2 = motifb$V2 + 1
  proximal = read.delim("proximal20.bed", stringsAsFactors = FALSE, header = FALSE)
  proximal$V2 = proximal$V2 + 1
  x2 = with(x, GRanges(chrom, IRanges(start, end), transcript_id = transcript_id))
  motifa = with(motifa, GRanges(V1, IRanges(V2, V3)))
 motifb = with(motifb, GRanges(V1, IRanges(V2, V3)))
 proximal = with(proximal, GRanges(V1, IRanges(V2, V3)))
 
 highlight3 = "Mutation6637696" #TERT
  highlight2 = c(x2[unique(queryHits(findOverlaps(x2, motifa)))]$transcript_id, x2[unique(queryHits(findOverlaps(x2, motifb)))]$transcript_id)
  highlight2 = unique(highlight2)
highlight2 = highlight2[highlight2 %in% highlight]
    
  highlight4 = unique(x2[unique(queryHits(findOverlaps(x2, proximal)))]$transcript_id)
  highlight4 = highlight4[highlight4 %in% highlight]
  
  par(new = TRUE)
  # Modified manhattan function
  chr = "CHR"
  bp = "bp"
  p = "pval"
  snp = "transcript_id"
  fdr = "fdr"
  col = c("gray9", "gray49")
  chrlabs = c(1:22, "X")
  logp = TRUE
  annotatePval = NULL
  annotateTop = TRUE
  div = 1
  
  CHR = BP = P = FDR = index = NULL
  
  if (!(chr %in% names(x))) stop(paste("Column", chr, "not found!"))
  if (!(bp %in% names(x))) stop(paste("Column", bp, "not found!"))
  if (!(p %in% names(x))) stop(paste("Column", p, "not found!"))
  if (!(fdr %in% names(x))) stop(paste("Column", fdr, "not found!"))
  if (!is.numeric(x[[chr]])) stop(paste(chr, "column should be numeric. Do you have 'X', 'Y', 'MT', etc? If so change to numbers"))
  if (!is.numeric(x[[bp]])) stop(paste(bp, "column should be numeric."))
  if (!is.numeric(x[[p]])) stop(paste(p, "column should be numeric."))
  if (!is.numeric(x[[fdr]])) stop(paste(fdr, "column should be numeric."))
  
  d = data.frame(transcript_id = x[["transcript_id"]], CHR = x[[chr]], BP = x[[bp]], P = x[[p]], FDR = x[[fdr]])
  
  if (!is.null(x[[snp]])) d = transform(d, SNP = x[[snp]])
  
  d <- subset(d, (is.numeric(CHR) & is.numeric(BP) & is.numeric(P) & is.numeric(FDR)))
  d <- d[order(d$CHR, d$BP), ]
  if (logp) {
    
    d$logp <- -log10(d$P)
    
  } else {
    
    d$logp <- d$P
  }
  
  d$pos = NA
  
  d$index = NA
  ind = 0
  for (i in unique(d$CHR)) {
    
    ind = ind + 1
    d[d$CHR == i, ]$index = ind
    
  }
  
  nchr = length(unique(d$CHR))
  if (nchr == 1) {
    
    d$pos = d$BP
    ticks = floor(length(d$pos)) / 2 + 1
    xlabel = paste('Chromosome', unique(d$CHR), 'position')
    labs = ticks
    
  } else { 
    
    lastbase = 0
    ticks = NULL
    for (i in unique(d$index)) {
      
      if (i == 1) {
        
        d[d$index == i, ]$pos = d[d$index == i, ]$BP
        
      } else {
        
        lastbase = lastbase + tail(subset(d, index == i - 1)$BP, 1)
        d[d$index == i, ]$pos = d[d$index == i, ]$BP + lastbase
        
      }
      ticks = c(ticks, (min(d[d$index == i, ]$pos) + max(d[d$index == i, ]$pos)) / 2 + 1)
      
    }
    xlabel = 'Chromosome'
    labs <- unique(d$CHR)
  }
  
  xmax = ceiling(max(d$pos) * 1.03)
  xmin = floor(max(d$pos) * (-0.03))
  
  plot(runif(10), runif(10), 
       xlim = c(xmin, xmax), ylim = c(0, ceiling(max(d$logp))), 
       axes = FALSE, # Don't plot the axis 
       type = "n",  # hide the points
       ylab = expression(-log[10](italic(p))), xlab = xlabel)
  
  axis(2, seq(0, ceiling(max(d$logp)), div))
  
  if (!is.null(chrlabs)) {
    
    if (is.character(chrlabs)) {
      
      if (length(chrlabs) == length(labs)) {
        
        labs <- chrlabs
        
      } else {
        
        warning("You're trying to specify chromosome labels but the number of labels != number of chromosomes.")
        
      }
      
    } else {
      
      warning ("If you're trying to specify chromosome labels, chrlabs must be a character vector")
      
    }
    
  }
  
  if (nchr == 1) {
    
    axis(1, at=ticks, labels = labs, las = 2, cex.axis = 0.7)
    
  } else {
    
    axis(1, at = ticks, labels = labs, las = 2, cex.axis = 0.7)
    
  }
  
  col = rep(col, max(d$CHR))
  
  if (nchr == 1) {
    
    with(d, points(pos, logp, pch = 16, col = col[1]))
    
  } else {
    
    icol = 1
    for (i in unique(d$index)) {
      
      with(d[d$index == unique(d$index)[i] & !d$transcript_id %in% c(highlight4), ], points(pos, logp, col = col[icol],  pch = 16))
      icol = icol + 1
      
    }
    
  }
  
  par(xpd = FALSE)
  
  # Highlight snps from a character vector
  if (!is.null(highlight)) {
    
    if (any(!(highlight %in% d$transcript_id))) warning("You're trying to highlight SNPs that don't exist in your results.")
    d.highlight = d[which(d$transcript_id %in% highlight & !d$transcript_id %in% highlight4), ]
    with(d.highlight, points(pos, logp, col = color.dots, pch = 16))
    
  }
  
  if (!is.null(fdr.cutoff)) {
    
    abline(h = (-log10(fdr.cutoff)), col = color.line)
    
  }
  
  # Highlight TERT
  if (!is.null(highlight3)) {
    
    if (any(!(highlight3 %in% d$transcript_id))) warning("You're trying to highlight SNPs that don't exist in your results.")
    d.highlight = d[which(d$transcript_id %in% highlight3), ]
    with(d.highlight, points(pos, logp, col = "red", pch = 16))
    
  }
  
  # Highlight proximal
  if (!is.null(highlight4)) {
    
    if (any(!(highlight4 %in% d$transcript_id))) warning("You're trying to highlight SNPs that don't exist in your results.")
    d.highlight = d[which(d$transcript_id %in% highlight4), ]
    with(d.highlight, points(pos, logp, col = color.dots, pch = 17))
    
  }
```

## Figure B
Comparisons among the hotspots detected by different models on melanoma samples
```{r, eval = FALSE}
# Naive - Simple binomial model
mask.regions.file =  mask.regions.file = system.file("extdata", "mask_regions.RDS", package = "MutSpot")
snv.mutations.file = snv.mutations.file2 = "melanoma_snv_mutations.MAF"
cores = 6
region.of.interest = NULL

chrOrder <- c(paste("chr", 1:22, sep = ""), "chrX")
seqi = GenomeInfoDb::seqinfo(BSgenome.Hsapiens.UCSC.hg19::Hsapiens)[GenomeInfoDb::seqnames(BSgenome.Hsapiens.UCSC.hg19::Hsapiens)[1:23]]
seqnames = GenomeInfoDb::seqnames(GenomeInfoDb::seqinfo(BSgenome.Hsapiens.UCSC.hg19::Hsapiens))[1:23]

mask.regions = readRDS(mask.regions.file)
mask.regions = mask.regions[as.character(GenomeInfoDb::seqnames(mask.regions)) %in% seqnames]

maf.snv <- maf.to.granges(snv.mutations.file)
maf.snv = maf.snv[as.character(GenomeInfoDb::seqnames(maf.snv)) %in% seqnames]
GenomeInfoDb::seqlevels(maf.snv) = as.character(unique(GenomeInfoDb::seqnames(maf.snv)))

maf.snv2 <- maf.to.granges(snv.mutations.file2)
maf.snv2 = maf.snv2[as.character(GenomeInfoDb::seqnames(maf.snv2)) %in% seqnames]
maf.ind.snv = GenomicRanges::split(maf.snv2, maf.snv2$sid)
ind.mut.count.snv = sapply(maf.ind.snv, length)

maf.masked.snv <- maf.snv[-S4Vectors::subjectHits(IRanges::findOverlaps(mask.regions, maf.snv))]
dupl.snv = duplicated(maf.masked.snv)
maf.uniq.snv = maf.masked.snv[!dupl.snv, ]

mut.regions = maf.uniq.snv + 15
names(mut.regions) = paste("Mutation", c(1:length(mut.regions)), sep = "")

rm(list = c("maf.uniq.snv", "mask.regions"))
gc()

genome.size = 2533374732
roi = mut.regions

print(paste("Using ", length(roi), " ROIs from input", sep = ""))

print(">> Intersecting ROIs and MAF ...")
maf.ovl <- IRanges::findOverlaps(roi, maf.snv, ignore.strand = TRUE)
maf.ovl.m = IRanges::as.matrix(maf.ovl)
min.count = 4

nsamples = length(unique(maf.snv$sid))
roi.count.mutated = tapply(maf.snv$sid[maf.ovl.m[ ,2]], names(roi)[maf.ovl.m[ ,1]],function(s) length(unique(s)))
roi.mutated = names(roi.count.mutated)[roi.count.mutated >= min.count]
print(paste("Using ", length(roi.mutated), " ROIs mutated in >=", min.count, " samples", sep = ""))

rois.to.run = roi.mutated

results = c()

sid=unique(maf.snv2$sid)

p.bg = parallel::mclapply (sid, function(s) {
  
  p = rep(length(maf.snv2) / (2533374732 * (length(unique(maf.snv2$sid)))), 21)
  p.roi.snv = 1 - prod(1 - p)   
  
  return(p.roi.snv)
  
}, mc.cores = 4 )
p.bg = unlist(p.bg)

process.single.roi <- function(x) {
  
  roi.progress <- which(rois.to.run == x)
  if ((roi.progress %% ceiling(length(rois.to.run) / 1000)) == 0) {
    
    print(paste("Progress : ", roi.progress, "/", length(rois.to.run), sep = ""))
    
  }
  
  x.idx = which(names(roi) == x)[1]
  k = length(unique(maf.snv$sid[maf.ovl.m[maf.ovl.m[ ,1] == x.idx, 2]]))
  
  pval = 1 - pbinom(k - 1, 70, unique(p.bg))
  
  c(x, pval, 31, mean(p.bg), k)
  
}

results = parallel::mclapply(rois.to.run, function(xr) { process.single.roi(xr) }, mc.cores = cores, mc.preschedule = FALSE, mc.silent = FALSE)

results.error = sapply(results, function(x) { length(x) == 1 } )
print(rois.to.run[which(results.error)])
print(results[results.error])
results = results[!results.error]

results = do.call(rbind,results)
rownames(results) = results[ ,1]
results = results[ ,-1, drop = FALSE]
class(results) <- "numeric"
fdr = p.adjust(results[ ,1], method = 'BH') * genome.size / nrow(results)
fdr = sapply(fdr, function(x) min(x, 1))  
results = cbind(results, fdr)
colnames(results) = c('pval', 'length', 'p.bg', 'k', 'fdr')
results = results[order(results[ ,'pval']), , drop = FALSE]

mut.rec = results
mut.regions2 = mut.regions[names(mut.regions) %in% rownames(mut.rec)]
mut.rec.hotspot = data.frame(chrom = as.character(GenomeInfoDb::seqnames(mut.regions2[rownames(mut.rec)])), start = GenomicRanges::start(mut.regions2[rownames(mut.rec)]), end = GenomicRanges::end(mut.regions2[rownames(mut.rec)]), mut.rec)

write.table(mut.rec.hotspot, file = "naive_hotspots_min4_31bp.tsv", sep = "\t", quote = FALSE, row.names = TRUE, col.names = TRUE)

mut.rec.hotspot2 = mut.rec.hotspot
mut.rec.hotspot2$ID = as.character(rownames(mut.rec.hotspot2))
mut.rec.hotspot2 = with(mut.rec.hotspot2, GenomicRanges::GRanges(chrom, IRanges::IRanges(start, end), pval = pval, length = length, p.bg = p.bg, k = k, fdr = fdr, ID = ID))
hotspots = IRanges::reduce(mut.rec.hotspot2)
hotspots$hs = paste("hs", 1:length(hotspots), sep = "")
ovl.mut = IRanges::findOverlaps(maf.snv, hotspots)
hotspots2 = hotspots[S4Vectors::subjectHits(ovl.mut)]
hotspots2$sample = maf.snv[S4Vectors::queryHits(ovl.mut)]$sid
hotspots2 = GenomicRanges::as.data.frame(hotspots2)
hotspots2 = aggregate(sample ~ hs, hotspots2, FUN = function(k) length(unique(k)))
colnames(hotspots2)[2] = "k"
rownames(hotspots2) = hotspots2$hs
hotspots$k = 0
for (i in 1:length(hotspots)) {
  # print(i)
  hotspots$k[i] = hotspots2[which(hotspots2$hs == hotspots$hs[i]), "k"]
  
}

ovl = IRanges::findOverlaps(mut.rec.hotspot2, hotspots)
mut.rec.hotspot2 = mut.rec.hotspot2[S4Vectors::queryHits(ovl)]
mut.rec.hotspot2$hs = hotspots[S4Vectors::subjectHits(ovl)]$hs
mut.rec.hotspot2$region.start = IRanges::start(hotspots[S4Vectors::subjectHits(ovl)])
mut.rec.hotspot2$region.end = IRanges::end(hotspots[S4Vectors::subjectHits(ovl)])
mut.rec.hotspot2$new.k = hotspots[S4Vectors::subjectHits(ovl)]$k
mut.rec.hotspot2 = GenomicRanges::as.data.frame(mut.rec.hotspot2)
mut.rec.hotspot2 = mut.rec.hotspot2[order(mut.rec.hotspot2$pval, decreasing = FALSE), ]
mut.rec.hotspot2 = mut.rec.hotspot2[!duplicated(mut.rec.hotspot2$hs), ]
mut.rec.hotspot2 = mut.rec.hotspot2[ ,c("seqnames", "region.start", "region.end", "pval", "length", "p.bg", "new.k", "fdr", "ID")]
mut.rec.hotspot2$length = mut.rec.hotspot2$region.end - mut.rec.hotspot2$region.start + 1
colnames(mut.rec.hotspot2) = c(colnames(mut.rec.hotspot), "ID")
rownames(mut.rec.hotspot2) = mut.rec.hotspot2$ID
mut.rec.hotspot2 = mut.rec.hotspot2[ ,-ncol(mut.rec.hotspot2)]

write.table(mut.rec.hotspot2, file = "naive_hotspots_31bp_min4_merged.tsv", sep = "\t", quote = FALSE, row.names = TRUE, col.names = TRUE)
```

```{r, eval = FALSE}
# MutSpot default model
MutSpot(snv.mutations = "melanoma_snv_mutations.MAF",
        cores = 6,
        hotspot.size = 31,
        fdr.cutoff = 0.01,
        min.count.snv = 4,
        drop = TRUE,
        cutoff.nucleotide.new = 1,
        genomic.features = "genomic_features_melanoma_default.txt")
```

```{r}
default = read.delim("snv_hotspots_merged_default_31bp_min4.tsv", stringsAsFactors = FALSE)
naive = read.delim("naive_hotspots_31bp_min4_merged.tsv", stringsAsFactors = FALSE)
context = read.delim("snv_hotspots_merged_melanoma_min4_31bp.tsv", stringsAsFactors = FALSE)

sum(naive$fdr <= 0.01) #1952
sum(default$fdr <= 0.01) #104
sum(context$fdr <= 0.01) #79

df = data.frame(models = c("naive", "default", "context"), hotspots = c(1952, 104, 79))
df$models = factor(df$models, levels = df$models)

#truncated yaxis
gap.barplot(y = c(1952, 104, 79), gap = c(from = 110, to = 1940), col = c("yellow", "yellow", "yellow"), 
            xlab = "index", ylab = "value", ytics = c(0, 50, 100, 1950))
axis.break(2, 110, breakcol = "snow", style = "gap")
axis.break(2, 110 * (1 + 0.02), breakcol = "black", style = "slash")
axis.break(4, 110 * (1 + 0.02), breakcol = "black", style = "slash")
axis(2, at = 110)
axis(2, at = c(0, 50, 100, 110), labels = NA) 
axis(4, at = c(0, 50, 100), labels = NA)
axis(2, at = c(0, 1960), labels = NA) 
```
